### PR TITLE
2334/fix date search in audit

### PR DESCRIPTION
**Pull request by plettich**
The DateTime objects from the database were not always converted to
ISO-strings properly. With custom dialect-based functions the formatting
is done in the SQL statement.

Working on #2334

### CODE REVIEW

File: privacyidea/lib/auditmodules/sqlaudit.py [hunk 1]
```diff
@@ -62,6 +64,32 @@
...
+# By using <https://docs.sqlalchemy.org/en/13/core/compiler.html> we can
+# differentiate between different dialects.
+class to_isodate(expression.FunctionElement):
+    name = 'to_isodate'
+
+
+@compiles(to_isodate, 'oracle')
+@compiles(to_isodate, 'postgresql')
+def fn_to_isodate(element, compiler, **kw):
+    return "to_char(%s, 'IYYY-MM-DD HH24:MI:SS')" % compiler.process(element.clauses, **kw)
```
Comment (cornelinux) [inline privacyidea/lib/auditmodules/sqlaudit.py:77]: I assume for whichever reason the "I" in "IYYY" and the "I" in "MI" is correct?

File: privacyidea/lib/auditmodules/sqlaudit.py [hunk 2]
```diff
@@ -188,9 +216,9 @@ def _create_filter(param, timelimit=None):
                     else:
                         # All other keys are compared as strings
                         column = getattr(LogEntry, search_key)
-                        if search_key == "date":
+                        if search_key.endswith("date"):
                             # but we cast "date" to a string first (required on postgresql)
```
Comment (cornelinux) [inline privacyidea/lib/auditmodules/sqlaudit.py:220]: Then the comment might be a bit different, now.

File: privacyidea/lib/auditmodules/sqlaudit.py [hunk 3]
```diff
@@ -188,9 +216,9 @@ def _create_filter(param, timelimit=None):
                     else:
                         # All other keys are compared as strings
                         column = getattr(LogEntry, search_key)
-                        if search_key == "date":
+                        if search_key.endswith("date"):
```
Comment (cornelinux) [inline privacyidea/lib/auditmodules/sqlaudit.py:219]: I hope we never have a column or searchkey like "mandate" ;-)

**Review by cornelinux — COMMENTED** [review]
I only have some minor comments. Please check if you think there is anything necessary from your side.

File: privacyidea/lib/auditmodules/sqlaudit.py (hunk 1 above)
Comment (plettich) [reply privacyidea/lib/auditmodules/sqlaudit.py:77]: These are the ISO format specifiers for Oracle/Postgres according to the respective docs:
https://docs.oracle.com/en/database/oracle/oracle-database/19/sqlrf/Format-Models.html#GUID-44E8F6D0-7532-4BE1-9300-F9775D9DB027
https://www.postgresql.org/docs/13/functions-formatting.html

File: privacyidea/lib/auditmodules/sqlaudit.py (hunk 2 above)
Comment (plettich) [reply privacyidea/lib/auditmodules/sqlaudit.py]: Done in fcff8608

File: privacyidea/lib/auditmodules/sqlaudit.py (hunk 3 above)
Comment (plettich) [reply privacyidea/lib/auditmodules/sqlaudit.py]: Done in fcff8608

**Review by cornelinux — APPROVED** [review]

**Pull request merged**